src/common/events.ts

```diff
@@ -0,0 +1,294 @@
+/**
+ * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { EventEmitter } from 'eventemitter3';
+import { PassThrough, Readable } from 'stream';
+
+import {
+  TurboSigningEmitterEventArgs,
+  TurboSigningEventsAndPayloads,
+  TurboUploadEmitterEventArgs,
+  TurboUploadEventsAndPayloads,
+} from '../types.js';
+
+/**
+ * Creates an eventing ReadableStream that emits progress and error events using the event names map.
+ *
+ * E.g.
+ *
+ * ```ts
+ * const eventNamesMap = {
+ *   'on-progress': 'signing-progress', // emits 'signing-progress' on event progress
+ *   'on-error': 'signing-error', // emits 'signing-error' errors
+ * };
+ *
+ * const streamWithEvents = createStreamWithEvents({
+ *   data,
+ *   dataSize,
+ *   emitter,
+ *   eventNamesMap,
+ * });
+ * ```
+ */
+function createReadableStreamWithEvents({
+  data,
+  dataSize,
+  emitter,
+  eventNamesMap,
+}: {
+  data: Buffer | ReadableStream;
+  dataSize: number;
+  emitter: EventEmitter;
+  eventNamesMap: {
+    'on-progress': string;
+    'on-error': string;
+  };
+}): ReadableStream {
+  const originalStream =
+    data instanceof ReadableStream
+      ? data
+      : new ReadableStream({
+          start: (controller) => {
+            controller.enqueue(data);
+            controller.close();
+          },
+        });
+
+  let processedBytes = 0;
+  let reader;
+  return new ReadableStream({
```

Comment:
Is it worth thinking about UX for attempting to read from a closed/finished stream?

Reply:
can you clarify?

Reply:
just thinking if there winds up being multiple references to the stream, how does it bubble up to the user. For example, if I am uploading a video, and i display a preview of the video in the upload UI, and then attempt to pass that video as a stream to the preview and the uploader, that would create two readers and one would finish before the other, throwing an error that the stream was already read, right?

Reply:
you're right to be concerned about how streams are consumed. that's why client interface requries providing `fileStreamFactory` or `dataItemStreamFactory` to tell the SDK "how do i create a new stream when I need it".

(separately, there is oppurtunity for us to use `.tee()` ReadableStreams and `PassThroughs` (Readables) and make those interfaces a bit easier to parse)